### PR TITLE
Chore update parser

### PR DESCRIPTION
**Pull request by evilebottnawi**
A lot of fixes.

Problem in `cast` and `retif` with `bin` nodes, i think problem in parser, so i will try to investigate and fix it tomorrow.

### CODE REVIEW

File: src/clean.js [hunk 1]
```diff
@@ -113,12 +113,9 @@ function clean(node, newObj) {
...
   }
 
-  // TODO: remove after resolve https://github.com/glayzzle/php-parser/issues/181
-  if (node.kind === "unary" && ["+", "-"].includes(node.type)) {
-    return {
-      kind: "number",
-      value: `${newObj.type}${newObj.what.value}`
-    };
+  // @TODO: We need special node for `null` value to avoid ast compare problem
+  if (node.kind === "classreference" && node.name.toLowerCase() === "null") {
```
Comment (evilebottnawi) [inline src/clean.js]: I think we can improve this in next `php-parser` version, right now it is normal solution.

File: src/comments.js [hunk 2]
```diff
@@ -283,6 +294,21 @@ function handleLastFunctionArgComments(
   return false;
 }
 
+function handleMemberExpressionComments(enclosingNode, followingNode, comment) {
```
Comment (evilebottnawi) [inline src/comments.js]: Now we handle comment in multiple property lookup:
```php
$fooBar
    ->doSomething('Hello World')
    ->doAnotherThing('Foo', ['foo' => $bar])

    // Now this comment in right place
    ->doOneMoreThing($config)

    ->run(function () {
        $console->log('Bar');
    });
```

File: src/comments.js [hunk 3]
```diff
@@ -520,7 +546,7 @@ function handleBreakAndContinueStatementComments(enclosingNode, comment) {
 }
 
 function handleGoto(enclosingNode, comment) {
-  if (enclosingNode && enclosingNode.kind === "goto") {
+  if (enclosingNode && ["label", "goto"].includes(enclosingNode.kind)) {
```
Comment (evilebottnawi) [inline src/comments.js]: Label now is identifier and can have comments

File: src/comments.js [hunk 4]
```diff
@@ -871,6 +894,121 @@ function canAttachComment(node) {
...
+      trailingParts.push(printTrailingComment(commentPath, print, options));
+    }
+  }, "comments");
+
+  return prependCursorPlaceholder(
+    path,
+    options,
+    concat(leadingParts.concat(trailingParts))
+  );
+}
```
Comment (evilebottnawi) [inline src/comments.js]: It is shame, but we need implement this api in prettier, right now i copy/paste code with little changes, i try to find on this issue in prettier in near future, now it is good solution

File: src/util.js [hunk 5]
```diff
@@ -310,13 +310,20 @@ function docShouldHaveTrailingNewline(path) {
...
 
-  if (["call", "new", "echo"].includes(parent.kind)) {
+  if (["call", "new"].includes(parent.kind)) {
     const lastIndex = parent.arguments.length - 1;
     const index = parent.arguments.indexOf(node);
 
     return index !== lastIndex;
   }
 
+  if (parent.kind === "echo") {
```
Comment (evilebottnawi) [inline src/util.js]: Now `echo` contains all nodes in `expressions` property

File: tests/comments/__snapshots__/jsfmt.spec.js.snap [hunk 6]
```diff
@@ -3184,7 +3185,7 @@ HERE;
 // Comment
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 <?php
-$str /* Comment */ = <<<EOD
+$str = /* Comment */ <<<EOD
```
Comment (evilebottnawi) [inline tests/comments/__snapshots__/jsfmt.spec.js.snap]: Awesome!

File: tests/comments/__snapshots__/jsfmt.spec.js.snap [hunk 7]
```diff
@@ -5143,9 +5144,12 @@ $a = new class() {
...
 $a = new Foo(/*Comment*/);
 $a = new Foo(/* Comment */ 1 /*Comments*/);
-$a = new Foo(); // Comment
+$a = new // Comment
+Foo();
 
-$a = new class {}; // Comment
+$a = new class// Comment
+ {
+};
```
Comment (evilebottnawi) [inline tests/comments/__snapshots__/jsfmt.spec.js.snap]: Issue already exists, let's fix it in other PR

File: tests/comments/__snapshots__/jsfmt.spec.js.snap [hunk 8]
```diff
@@ -5554,7 +5558,8 @@ echo $a/*Comment*/->/*Comment*/bar();
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 <?php
 
-echo /*Comment*/ /*Comment*/ $a->bar();
+echo $a /*Comment*/
+    ->/*Comment*/ bar();
```
Comment (evilebottnawi) [inline tests/comments/__snapshots__/jsfmt.spec.js.snap]: Let's solve this in other PR

Comment (evilebottnawi) [reply tests/comments/__snapshots__/jsfmt.spec.js.snap]: https://github.com/prettier/plugin-php/issues/870

File: tests/comments/__snapshots__/jsfmt.spec.js.snap [hunk 9]
```diff
@@ -5644,9 +5649,8 @@ $extractTextPluginOptions = $shouldUseRelativeAssetPaths // Making sure that the
     : [];
 
 $var =
-    $process->env->NODE_ENV ===
-    "production" // a
-        ? call("./configureProdStore")
+    $process->env->NODE_ENV === "production"
+        ? call("./configureProdStore") // a
```
Comment (evilebottnawi) [inline tests/comments/__snapshots__/jsfmt.spec.js.snap]: Awesome!

File: tests/comments/__snapshots__/jsfmt.spec.js.snap [hunk 10]
```diff
@@ -5767,7 +5771,7 @@ echo Foo/*Comment*/::/*Comment*/bar();
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 <?php
 
-echo /*Comment*/ /*Comment*/ Foo::bar();
+echo /*Comment*/ Foo::/*Comment*/ bar();
```
Comment (evilebottnawi) [inline tests/comments/__snapshots__/jsfmt.spec.js.snap]: Better

File: tests/comments/__snapshots__/jsfmt.spec.js.snap [hunk 11]
```diff
@@ -5910,12 +5914,12 @@ trait tFoo {
 }
 
 // Comment
-trait tFoo extends /* Comment */ ReflectionMethod {}
+trait tFoo /* Comment */ {}
```
Comment (evilebottnawi) [inline tests/comments/__snapshots__/jsfmt.spec.js.snap]: traits don't support `extends` and `implements` and no parser throw error on this, so i remove this

File: tests/comments/__snapshots__/jsfmt.spec.js.snap [hunk 12]
```diff
@@ -5956,24 +5960,22 @@ trait tFoo
...
 }
 
 class ImplementingClass
 {
     /* Comment */
-    /* Comment */ /* Comment */ use testTrait /* Comment */,
+    /* Comment */ use /* Comment */ testTrait /* Comment */,
         /* Comment */ otherTrait /* Comment */;
-    /* Comment */ /* Comment */ use testTrait /* Comment */,
+    /* Comment */ use /* Comment */ testTrait /* Comment */,
```
Comment (evilebottnawi) [inline tests/comments/__snapshots__/jsfmt.spec.js.snap]: Expected after fixing comment positions

File: tests/encapsed/__snapshots__/jsfmt.spec.js.snap [hunk 13]
```diff
@@ -546,12 +546,12 @@ $var = "string $var[0][0] string";
...
-$var = "string \${$var->foo} string";
-$var = "string \${$var->call()} string";
-$var = "string \${$var::$var} string";
-$var = "string \${$var::call()} string";
+$var = "string {\${$var}} string";
+$var = "string {\${call()}} string";
+$var = "string {\${$var->foo}} string";
+$var = "string {\${$var->call()}} string";
+$var = "string {\${$var::$var}} string";
+$var = "string {\${$var::call()}} string";
```
Comment (evilebottnawi) [inline tests/encapsed/__snapshots__/jsfmt.spec.js.snap]: Now we respect `{` and `}` in `encapsed` and it is awesome

File: tests/preserve_line/__snapshots__/jsfmt.spec.js.snap [hunk 14]
```diff
@@ -492,50 +492,49 @@ $sel = $this->connections
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 <?php
 
-// App configuration.
 $fooBar
     ->doSomething('Hello World')
     ->doAnotherThing('Foo', ['foo' => $bar])
 
+    // App configuration.
     ->doOneMoreThing($config)
```
Comment (evilebottnawi) [inline tests/preserve_line/__snapshots__/jsfmt.spec.js.snap]: As i write above, now we respect comment in this places

File: tests/staticlookup/__snapshots__/jsfmt.spec.js.snap [hunk 15]
```diff
@@ -260,19 +260,16 @@ $var = $this->modelClass::where('name', 'like', strtoupper("%\${key}%"))->get();
...
+        $arrayStub->baz
+    ];
 $var =
-    parent
-        ::$arrayCache[$arrayStub->class][$arrayStub->position][$arrayStub->foo][
-            $arrayStub->baz
-        ];
+    parent::$arrayCache[$arrayStub->class][$arrayStub->position][
+        $arrayStub->foo
+    ][$arrayStub->baz];
```
Comment (evilebottnawi) [inline tests/staticlookup/__snapshots__/jsfmt.spec.js.snap]: Now we better print `static`/`self`/`parent` chains

**Comment by evilebottnawi** [comment]
Also need implement more test, let's do it in other PRs

**Comment by evilebottnawi** [comment]
I think we can add `--ignore-engines` as temprorary solution to avoid problem in `jsdom` (https://github.com/prettier/plugin-php/pull/868)

File: src/comments.js [hunk 16]
```diff
@@ -5,11 +5,21 @@ const {
...
+  join,
+  indent,
+  hardline,
+  cursor,
+  lineSuffix,
+  breakParent
+} = require("prettier").doc.builders;
 // TODO: remove after resolve https://github.com/prettier/prettier/pull/5049
-const { hasNewline, hasNewlineInRange } = require("./util");
+const { hasNewline, hasNewlineInRange, isLookupNode } = require("./util");
```
Comment (czosel) [inline src/comments.js]: Let's take this out of `prettier.util` once #868 is merged

Comment (evilebottnawi) [reply src/comments.js]: :+1: Yep, i think we can implement and fix all problem in this/next week and release new version in next week too, (Fix unnecessary todo in my todo list :smile: )

Comment (czosel) [reply src/comments.js]: Awesome, can't wait! :tada:

**Pull request merged**